rasa/core/policies/ted_policy.py

```diff
@@ -755,39 +812,110 @@ def _emebed_dialogue(
             A tensor combining  all features for `attribute`
         """
 
-        if not tf_batch_data[attribute]:
-            return None
-
         attribute_mask = tf_batch_data[attribute][MASK][0]
-        # TODO transformer has to be used to process sequence features
-        attribute_features = self._combine_sparse_dense_features(
-            tf_batch_data[attribute][SENTENCE],
-            f"{attribute}_{SENTENCE}",
-            mask=attribute_mask,
-        )
 
-        if attribute in FEATURES_TO_ENCODE + LABEL_FEATURES_TO_ENCODE:
-            attribute_features = self._tf_layers[f"ffnn.{attribute}_{SENTENCE}"](
+        if attribute in SEQUENCE_FEATURES_TO_ENCODE:
+            _sequence_lengths = tf.cast(
+                tf_batch_data[attribute][SEQUENCE_LENGTH][0], dtype=tf.int32
+            )
+            mask_sequence_text = self._compute_mask(tf.squeeze(_sequence_lengths))
+            sequence_lengths = _sequence_lengths + 1
+            mask_text = self._compute_mask(tf.squeeze(sequence_lengths))
+
+            attribute_features, _, _, _ = self._create_sequence(
+                tf_batch_data[attribute][SEQUENCE],
+                tf_batch_data[attribute][SENTENCE],
+                mask_sequence_text,
+                mask_text,
+                attribute,
+                sparse_dropout=self.config[SPARSE_INPUT_DROPOUT],
+                dense_dropout=self.config[DENSE_INPUT_DROPOUT],
+                masked_lm_loss=self.config[MASKED_LM],
+                sequence_ids=False,
+            )
+
+            # TODO entities
+
+            # resulting attribute features will have shape
+            # combined batch dimension and dialogue length x sequence length x units
+            attribute_features = self._last_token(
+                attribute_features, tf.squeeze(sequence_lengths)
+            )
+
+        else:
+            # resulting attribute features will have shape
+            # combined batch dimension and dialogue length x 1 x units
+            attribute_features = self._combine_sparse_dense_features(
+                tf_batch_data[attribute][SENTENCE],
+                f"{attribute}_{SENTENCE}",
+                mask=attribute_mask,
+            )
+
+        if attribute in SENTENCE_FEATURES_TO_ENCODE + LABEL_FEATURES_TO_ENCODE:
+            attribute_features = self._tf_layers[f"ffnn.{attribute}"](
                 attribute_features
             )
 
+        if attribute in SENTENCE_FEATURES_TO_ENCODE:
+            dialogue_lengths = tf.cast(
+                tf_batch_data[DIALOGUE][f"3D_{LENGTH}"][0], tf.int32
+            )
+
+            attribute_features = self._convert_to_original_shape(
+                attribute_features, dialogue_lengths
+            )
+
+            # create a attribute mask that has the shape
+            # batch x dialogue length
+            attribute_mask = tf.expand_dims(
+                tf.squeeze(self._compute_mask(tf.squeeze(dialogue_lengths))), axis=-1
+            )
```

Comment:
`attribute_mask` is a different mask, it is not about dialogue lengths, but it tracks whether the input is fake input (zero features) or real input. E.g. where there is intent, there is no text and vice versa

Reply:
that is actually exactly what `dialogue_length` is storing in the 3D way... it is 32 x 27 x 1, with just 1 or 0 depending on whether the dialogue turn is present or not. should we maybe rename it?

Reply:
but it's not about the dialogue turn, attribute mask stores whether the certain attribute is present or it is the fake one

Reply:
just realized that dialogue length is not attribute specific. so it just indicates in general if we have any actual (aka non zero) features for this dialogue or not. I think we actually need the mask per attribute... we might need to add more info to the model data in that case.

Reply:
found a much simpler way... we can just reuse the "4D" attribute mask